Notes/Algorithms/Data Structures/Hash-Tables.c

```diff
@@ -1,37 +1,345 @@
+        return ERROR;
+    }
+}
+
+
+// testInsertCapcity
+// prints an error message and returns ERROR if insert is unsuccessful due to hashTable being at capcity
+// returns SUCCESS if all key/value pairs can be inserted into the hashTable
+int testInsertCapacity() {
+    HashTable * testHash = createHashTable(10);
+    char keys[100][10] = {
+        "KeyOne", "KeyTwo", "KeyThree", "KeyFour", "KeyFive", "KeySix", "KeySeven", "KeyEight", "KeyNine", "KeyTen",
+        "KeyOne", "KeyTwo", "KeyThree", "KeyFour", "KeyFive", "KeySix", "KeySeven", "KeyEight", "KeyNine", "KeyTen",
+        "KeyOne", "KeyTwo", "KeyThree", "KeyFour", "KeyFive", "KeySix", "KeySeven", "KeyEight", "KeyNine", "KeyTen",
+        "KeyOne", "KeyTwo", "KeyThree", "KeyFour", "KeyFive", "KeySix", "KeySeven", "KeyEight", "KeyNine", "KeyTen",
+        "KeyOne", "KeyTwo", "KeyThree", "KeyFour", "KeyFive", "KeySix", "KeySeven", "KeyEight", "KeyNine", "KeyTen",
+        "KeyOne", "KeyTwo", "KeyThree", "KeyFour", "KeyFive", "KeySix", "KeySeven", "KeyEight", "KeyNine", "KeyTen",
+        "KeyOne", "KeyTwo", "KeyThree", "KeyFour", "KeyFive", "KeySix", "KeySeven", "KeyEight", "KeyNine", "KeyTen",
+        "KeyOne", "KeyTwo", "KeyThree", "KeyFour", "KeyFive", "KeySix", "KeySeven", "KeyEight", "KeyNine", "KeyTen",
+        "KeyOne", "KeyTwo", "KeyThree", "KeyFour", "KeyFive", "KeySix", "KeySeven", "KeyEight", "KeyNine", "KeyTen",
+        "KeyOne", "KeyTwo", "KeyThree", "KeyFour", "KeyFive", "KeySix", "KeySeven", "KeyEight", "KeyNine", "KeyTen",
+    };
+    int values[100] = {
+         10, 20, 30, 40, 50, 60, 70, 80, 90, 100,
+         10, 20, 30, 40, 50, 60, 70, 80, 90, 100,
+         10, 20, 30, 40, 50, 60, 70, 80, 90, 100,
+         10, 20, 30, 40, 50, 60, 70, 80, 90, 100,
+         10, 20, 30, 40, 50, 60, 70, 80, 90, 100,
+         10, 20, 30, 40, 50, 60, 70, 80, 90, 100,
+         10, 20, 30, 40, 50, 60, 70, 80, 90, 100,
+         10, 20, 30, 40, 50, 60, 70, 80, 90, 100,
+         10, 20, 30, 40, 50, 60, 70, 80, 90, 100,
+         10, 20, 30, 40, 50, 60, 70, 80, 90, 100
+    };
+    int success = SUCCESS;
+    for (int i = 0; i < 100; i++) {
+        success = insert(testHash, keys[i], values[i]);
+        if (success != SUCCESS) {
+            printf("ERROR! HashTable is at capacity! No more values can be added. \n");
+            return ERROR;
+        }
+    }
+    return SUCCESS;
+
+}
+
+// buildInsertTests
+// creates a test for the case of collisions
+// createas a test for the case of hashTable capcity being reached
+void buildInsertTests() {
+   //build an empty hashTable to test insert/collision on
+    HashTable * testHash = createHashTable(10);
+    printf("Build Insert Collision Test: \n");
+
+   // Brute Forced a Collision (Same Key = Same hashingIndex)
+   // Expect that the second KeyOne will have an index = hashing(KeyOne) + 1 % hashTable->size
+    testInsertCollision(testHash, "KeyOne", 10, "KeyOne", 20);
```

Comment:
This is an interesting illustration of a design decision you made.  Typically, when given a duplicate key, a hash table will update the value for the extant item in place.  Your implementation currently treats the second insert as a unique item.

As a thought (or code) experiment, what is the behavior of your table given the following pseudocode input series:

```
h = createHashTable(10)
insert(h, "foo", 1) // stores in slot i
insert(h, "foo", 2)  // stores in slot i + 1
delete(h, "foo")  // deletes slot i
delete(h, "foo") // nothing in slot i?  Is it possible to delete the item in i + 1?
```
For other testing purposes, here are some collisions that aren't the same word:
```
$ cat loselose.py 
def loselose(s):
    h = 0
    for c in s:
        h += ord(c)
    return h

if __name__ == "__main__":
    map = {}
    with open("/usr/share/dict/words", 'r') as fh:
        for line in fh:
            word = line.strip()
            h = loselose(word)
            if not h in map:
                map[h] = [word]
            else:
                map[h].append(word)
    for hash, wordlist in map.items():
        if len(wordlist) > 1:
            print("%i collisions: %s" % (hash, ", ".join(wordlist)))

$ python loselose.py | grep '^450'
450 collisions: Kit's, Lew's, Mir's, bury, cad's, crux, drys, envy, lost, lots, moss, myth, only, plow, pops, pump, ruby, rush, sews, slot, stir, twin, type, upon, vary, vest, vets, whiz, zany
```

Reply:
Big update in [commit 5abee1041869a8a7b2fc48bf295fd7fef01e9470](https://github.com/kcarrel/Training-Notes/pull/12/commits/5abee1041869a8a7b2fc48bf295fd7fef01e9470).

Note: in writing this comment response I came to the conclusion that replace was a better [function name](https://github.com/kcarrel/Training-Notes/pull/12/commits/6309892df7c3151eb76a16199422ab751377eb9c).

Changed the design of insert to handle duplicate keys by finding the key in the hashTable (accounting for potential that the key is not at the initially calculated hashIndex but may have experienced a collision), delete the current key/value pair once found then replace with the new key/value pair.

Added in a helper function in the case that there is a duplicate key to replace but it is not at the expected initially calculate hashIndex.

Added in a test to check insert against a duplicate key scenario.

Changed the Insert Collision test to take in two key/value pairs that are brute forced collisions without same key.

Explicitly declared searchHashTable and delete so that it can be used by insert and replace functions. (Note: Would love feedback/thoughts on if using an mix of explicit and implicit declaration of functions throughout code is acceptable - Normally I would go for one consistent style but thought that moving around large chunks of code at this point and for the purposes of this to maintain implicit declaration consistency would be more of a problem than a solution.)